### PR TITLE
Add feature flag db.enable_paginate_window

### PR DESCRIPTION
**Pull request by joaopapereira**
Thanks for contributing to cloud_controller_ng. To speed up the process of reviewing your pull request, please provide us with the following:

* A short explanation of the proposed change:
This feature flag, if not present, will default to the current behavior of using the window pagination, but when it is set to false, it will fall back on using the other pagination method when accessing the database.

* An explanation of the use cases your change solves
This change will allow the operator to disable the window for pagination for all the tables in a single place. Will also allow us to debug possible performance issues with the window pagination in other tables.

* Links to any other associated PRs

* [ x ] I have reviewed the [contributing guide](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/CONTRIBUTING.md)

* [ x ] I have viewed, signed, and submitted the Contributor License Agreement

* [ x ] I have made this pull request to the `main` branch

* [ x ] I have run all the unit tests using `bundle exec rake`

* [  ] I have run [CF Acceptance Tests](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/spec/README.md#cf-acceptance-tests-cats)

### CODE REVIEW

File: lib/cloud_controller/paging/sequel_paginator.rb [hunk 1]
```diff
@@ -25,7 +25,10 @@ def get_page(dataset, pagination_options)
     end
 
     def can_paginate_with_window_function?(dataset)
-      dataset.supports_window_functions? && (!dataset.opts[:distinct] || !dataset.requires_unique_column_names_in_subquery_select_list?)
+      enable_paginate_window = true if Config.config.get(:db, :enable_paginate_window).nil?
+      enable_paginate_window = Config.config.get(:db, :enable_paginate_window) unless Config.config.get(:db, :enable_paginate_window).nil?
```
Comment (philippthun) [inline lib/cloud_controller/paging/sequel_paginator.rb:29]: Proposal for a shorter and more concise reading of the config value:
```
enable_paginate_window = Config.config.get(:db, :enable_paginate_window).nil? ? true : Config.config.get(:db, :enable_paginate_window)
```

File: spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb [hunk 2]
```diff
@@ -204,6 +204,77 @@ module VCAP::CloudController
           expect(paginated_result.total).to be > 1
         end
       end
+
+      context 'window controll config flag' do
```
Comment (philippthun) [inline spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb:208]:
```suggestion
      context 'enable_paginate_window config flag' do
```

File: spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb [hunk 3]
```diff
@@ -204,6 +204,77 @@ module VCAP::CloudController
...
+        let(:dataset) { AppModel.dataset }
+        let!(:app_1) { AppModel.make(guid: '1', created_at: '2024-05-15T17:23:01Z') }
+        let!(:app_2) { AppModel.make(guid: '2', created_at: '2024-05-15T17:23:02Z') }
+        let!(:app_3) { AppModel.make(guid: '3', created_at: '2024-05-15T17:23:03Z') }
+        let!(:app_4) { AppModel.make(guid: '4', created_at: '2024-05-15T17:23:04Z') }
+
+        context 'not defined' do
+          it 'uses window function' do
```
Comment (philippthun) [inline spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb:216]:
```suggestion
          it 'uses window function if supported' do
```

File: spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb [hunk 4]
```diff
@@ -204,6 +204,77 @@ module VCAP::CloudController
...
+            options = { page:, per_page: }
+            pagination_options = PaginationOptions.new(options)
+
+            paginated_result = nil
+            expect do
+              paginated_result = paginator.get_page(dataset, pagination_options)
+            end.to have_queried_db_times(/select/i, 1)
```
Comment (philippthun) [inline spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb:223]:
```suggestion
            end.to have_queried_db_times(/select/i, dataset.supports_window_functions? ? 1 : 2)
```

File: spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb [hunk 5]
```diff
@@ -204,6 +204,77 @@ module VCAP::CloudController
...
+                enable_paginate_window: true
+              }
+            }
+          end
+
+          before do
+            TestConfig.override(**my_config)
+          end
+
+          it 'uses window function' do
```
Comment (philippthun) [inline spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb:241]:
```suggestion
          it 'uses window function if supported' do
```

File: spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb [hunk 6]
```diff
@@ -204,6 +204,77 @@ module VCAP::CloudController
...
+            options = { page:, per_page: }
+            pagination_options = PaginationOptions.new(options)
+
+            paginated_result = nil
+            expect do
+              paginated_result = paginator.get_page(dataset, pagination_options)
+            end.to have_queried_db_times(/select/i, 1)
```
Comment (philippthun) [inline spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb:248]:
```suggestion
            end.to have_queried_db_times(/select/i, dataset.supports_window_functions? ? 1 : 2)
```

File: spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb [hunk 7]
```diff
@@ -204,6 +204,77 @@ module VCAP::CloudController
...
+                enable_paginate_window: false
+              }
+            }
+          end
+
+          before do
+            TestConfig.override(**my_config)
+          end
+
+          it 'uses window function' do
```
Comment (philippthun) [inline spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb:266]:
```suggestion
          it 'does not use window function' do
```

File: spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb (hunk 4 above)
Comment (philippthun) [reply spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb:223]: The test currently fails for those databases that do not support window functions. There are other tests in this file where we also have different expectations depending on the db's capabilities.

File: lib/cloud_controller/paging/sequel_paginator.rb (hunk 1 above)
Comment (joaopapereira) [reply lib/cloud_controller/paging/sequel_paginator.rb]: Fixed

File: spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb (hunk 3 above)
Comment (joaopapereira) [reply spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb]: Fixed

File: spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb (hunk 5 above)
Comment (joaopapereira) [reply spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb]: Fixed

File: spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb (hunk 2 above)
Comment (joaopapereira) [reply spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb]: Fixed

File: spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb (hunk 4 above)
Comment (joaopapereira) [reply spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb]: Fixed

File: spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb (hunk 6 above)
Comment (joaopapereira) [reply spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb]: Fixed

File: spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb (hunk 7 above)
Comment (joaopapereira) [reply spec/unit/lib/cloud_controller/paging/sequel_paginator_spec.rb]: Fixed

**Review by philippthun — APPROVED** [review]

**Pull request merged**